### Guided Project: 
# Exploring eBay Car Sales Data

The aim of this project is to clean 370k+ rows x 20 columns amount of eBay car sales data and analyze the included used car listings. You'll also become familiar with some of the unique benefits *Jupyter Notebook* provides for *Pandas.* 

You could find the data [here](https://data.world/data-society/used-cars-data) (originally from Kaggle by user orgesleka)

In [1]:
# Import Pandas and NumPy
import pandas as pd
import numpy as np

# Read the data (pd)
autos = pd.read_csv('autos.csv', encoding='Latin-1')

# Check the data info
autos.info()

# Check first 5 data
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Data Cleaning: Columns

You've noticed that several data are missing in the `info`, they include `vehicleType`, `gearbox`, `model`, `fuelType`, and `notRepairedDamage`. We can clean these data in order to analyze it well. Also, the writing format use "CamelCase", where you connect the words with capital letters. The convention is to use "snake_case", where each word is connected by character underscore ("\_"). We can tackle this one first.

In `DataFrame.info()`, we can see 20 columns data with its respective type. The 5 first data can be portrayed through `DataFrame.head()`

In [2]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


`# Rename the CamelCase columns to snake_case`
`# FIRST TRY:`
`autos.rename(columns={'dateCrawled': 'date_crawled'}, inplace=True)`
`autos.rename(columns={'offerType': 'offer_type'}, inplace=True)`
`autos.rename(columns={'vehicleType': 'vehicle_type'}, inplace=True)`
`autos.rename(columns={'yearOfRegistration': 'year_of_registration'}, inplace=True)`
`autos.rename(columns={'powerPS': 'power_ps'}, inplace=True)`
`autos.rename(columns={'monthOfRegistration': 'month_of_registration'}, inplace=True)`
`autos.rename(columns={'fuelType': 'fuel_type'}, inplace=True)`
`autos.rename(columns={'notRepairedDamage': 'not_repaired_damage'}, inplace=True)`
`autos.rename(columns={'dateCreated': 'date_created'}, inplace=True)`
`autos.rename(columns={'nrOfPictures': 'nr_of_pictures'}, inplace=True)`
`autos.rename(columns={'postalCode': 'postal_code'}, inplace=True)`
`autos.rename(columns={'lastSeen': 'last_seen'}, inplace=True)`

^ First try -- Not effective.
We can change it in a line instead like below:

In [3]:
# Changing all columns in one line.
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 
                 'price', 'abtest', 'vehicle_type', 'reg_year',
                 'gearbox', 'power_ps', 'model', 'odometer',
                 'reg_month', 'fuel_type', 'brand', 
                 'unrepaired_damage', 'ad_created', 'nr_of_pictures',
                 'postal_code', 'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  reg_year    gearbox  \
0  privat    Angebot  $5,000  control          bus      2004    manuell   
1  privat    Angebot  $8,500  control    limousine      1997  automatik   
2  privat    Angebot  $8,990     test    limousine      2009    manuell   
3  privat    Angebot  $4,350  control   kleinwagen      2007  automatik   
4  privat    Angebot  $1,350     test        kombi      2003    manuell   

   power_ps   model   odometer  reg_month fuel_type       brand  \
0       158  andere  150,000km          3       lpg     peugeot   
1       286     7er  150,000km          6    benzin         bmw   
2       102    golf   70,000km          7    benzin  volkswagen   
3        71  fortwo   70,000km          6    benzin       smart   
4         0   focus  150,000km          7    benzin        ford   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   
3              nein  2016-03-12 00:00:00               0        33729   
4              nein  2016-04-01 00:00:00               0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

There we go :D

We can also change the data format and type of both `price` and `odometer`. As shown in autos.info(), they're in `str` format, while we want to calculate them with the `int` type.

In [4]:
# We can check the excel for some skimming, or..

# we can do `DataFrame.describe() to pinpoint out
# data that can be deleted for efficiency
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type      reg_year  gearbox      power_ps  model   odometer  \
count         44905  50000.000000    47320  50000.000000  47242      50000   
unique            8           NaN        2           NaN    245         13   
top       limousine           NaN  manuell           NaN   golf  150,000km   
freq          12859           NaN    36993           NaN   4024      32424   
mean            NaN   2005.073280      NaN    116.355920    NaN        NaN   
std             NaN    105.712813      NaN    209.216627    NaN        NaN   
min             NaN   1000.000000      NaN      0.000000    NaN        NaN   
25%             NaN   1999.000000      NaN     70.000000    NaN        NaN   
50%             NaN   2003.000000      NaN    105.000000    NaN        NaN   
75%             NaN   2008.000000      NaN    150.000000    NaN        NaN   
max             NaN   9999.000000      NaN  17700.000000    NaN        NaN   

           reg_month fuel_type       brand unrepaired_damage  \
count   50000.000000     45518       50000             40171   
unique           NaN         7          40                 2   
top              NaN    benzin  volkswagen              nein   
freq             NaN     30107       10687             35232   
mean        5.723360       NaN         NaN               NaN   
std         3.711984       NaN         NaN               NaN   
min         0.000000       NaN         NaN               NaN   
25%         3.000000       NaN         NaN               NaN   
50%         6.000000       NaN         NaN               NaN   
75%         9.000000       NaN         NaN               NaN   
max        12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The columns of with less `unique` may indicate the lack of data. This also applies to columns whose `mean` to `max` equals 0. Let's check it out!

In [5]:
## Columns whose `unique` values are only 1/2
print(autos["seller"].value_counts())
print(autos["offer_type"].value_counts())
print(autos["abtest"].value_counts())
print(autos["gearbox"].value_counts())
print(autos["unrepaired_damage"].value_counts())

## Columns whose `mean` to `max` data equals 0
print(autos["nr_of_pictures"].value_counts())

privat        49999
gewerblich        1
Name: seller, dtype: int64
Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
test       25756
control    24244
Name: abtest, dtype: int64
manuell      36993
automatik    10327
Name: gearbox, dtype: int64
nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64
0    50000
Name: nr_of_pictures, dtype: int64


As you can see, columns `seller` and `offer_type` has nearly all the same values. Also, the column `nr_of_picture` does not have any data whatsoever. Sooo, we're deleting these columns.

In [6]:
# Drop these columns in axis=1
autos = autos.drop(["seller", "offer_type", "nr_of_pictures"], axis=1)
print(autos.head())

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  reg_year    gearbox  power_ps   model  \
0  $5,000  control          bus      2004    manuell       158  andere   
1  $8,500  control    limousine      1997  automatik       286     7er   
2  $8,990     test    limousine      2009    manuell       102    golf   
3  $4,350  control   kleinwagen      2007  automatik        71  fortwo   
4  $1,350     test        kombi      2003    manuell         0   focus   

    odometer  reg_month fuel_type       brand unrepaired_damage  \
0  150,000km          3  

In [7]:
# Clean the "$" in `price` to have the number data only, 
# and change `price` type from `object` or `str` to `int`.
autos["price"] = (autos["price"]
                            .str.replace("$", "")
                            .str.replace(",", "")
                            .astype(int)
                 )

# Change data format in `odometer` from `10000km` to `10000`,
# and then change the type into `int` and rename it `odometer_km`
autos["odometer"] = (autos["odometer"]
                                     .str.replace("km", "")
                                     .str.replace(",", "")
                                     .astype(int)
                    )
                     
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

Let's check if there are outliers in both data!

In [8]:
uniq_price = autos["price"].unique().shape[0]
desc_price = autos["price"].describe()

# One thing I learned:
# As long as the output remain the same as input
# (for instance Series input to Series output),
# you can chain the functions aaaall at once.

valc_price_desc = (autos["price"]
                          .value_counts()
                          .sort_index(ascending=False)
                          .head(40)
                 )

valc_price_asc = (autos["price"]
                          .value_counts()
                          .sort_index(ascending=True)
                          .head(20)
                 )

print(uniq_price, "\n")
print(desc_price, "\n")
print(valc_price_desc, "\n")
print(valc_price_asc, "\n")

2357 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
Name: price, dtype: int64 

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
3

There's a spectrum of outliers around the data. We can clean it out by putting ranges in between. And to make it easier, we can use the built-in method `Series.between(x,y)` as the argument.

In [13]:
# Spectrum range: 1 - 351000
autos = autos[ (autos["price"].between(1,351000)) ]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

We've cleaned the data from 50k to 48.5k with `price` range of 1 - 350,000 dollars. \[WITH HELP\] Lesson learned: You need to recheck what's the range gonna be by sorting it in Excel, that way you can find out organically. Based on the sorting, `price` above 350000 dollars tend to skyrocket to an outlier `price`. That way we can clean it out.

Let's check out the date format!

## Data Cleaning: Dates

In [10]:
autos_date = autos[['date_crawled', 'ad_created', 'last_seen']]
print(autos_date[0:5])

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50


In [15]:
## LESSON LEARNED:
# You can print the first several letters of `str`
# in each series/dataframe. That way you can pickup
# what's date and what's time. Let's look at code below.

print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


We can sort the dates by (1) segmenting the dates or 10 first letters, (2) sort it out with built-in function `sorted_values()`, with addition of arguments `normalize=True` for converting the values to percentages and `dropna=False` to not drop the NaN values. We do this for each `Series` (There's no method like this in `Dataframe`, so you gotta do it one-by-one)

In [18]:
## `date_crawled`
print(autos["date_crawled"]
      .str[:10]
      .value_counts(
          normalize=True, 
          dropna=False)
      .sort_index(
          ascending=True)
     )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


In [19]:
## `ad_created`
print(autos["ad_created"]
      .str[:10]
      .value_counts(
          normalize=True, 
          dropna=False)
      .sort_index(
          ascending=True)
     )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64


In [20]:
## `last_seen`
print(autos["last_seen"]
      .str[:10]
      .value_counts(
          normalize=True, 
          dropna=False)
      .sort_index(
          ascending=True)
     )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


We can also see the distribution in year of registration `reg_year` to see the car-buying trend.

In [22]:
## reg_year

print(autos["reg_year"]
      .value_counts(
          normalize=True, 
          dropna=False)
      .sort_index(
          ascending=True)
     )

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: reg_year, Length: 95, dtype: float64


$New Discovery:$ We've noticed that there are several data whose years in `year` are invalid, i.e. 1000 (way in the past) and 9999 (way in the future). We can clean them out using the same code with `between(x,y)`.

To decide the range, above 2016 is not accurate, since the data is handled at that time. We can take 1900 as the year car invented. Or no. Actually, let's see when eBay started. $*searching*$ 1995, that's reasonable. So, 1995-2016 it is.

In [25]:
autos = autos[ (autos["reg_year"].between(1995,2016)) ]
autos["reg_year"].describe()
autos["reg_year"].value_counts(normalize=True).head(10)

2000    0.072931
2005    0.067847
1999    0.066946
2004    0.062462
2003    0.062370
2006    0.061700
2001    0.060914
2002    0.057448
1998    0.054606
2007    0.052618
Name: reg_year, dtype: float64

Now, we've cleaned the data further from 48,5k to 43,3k. That's neat :D

## Exploring Data: Price by Brand
Now seeing that the data's clean, we can move on to exploration!

We can look for the popular brand bought over the years by aggregating the `brand` data into `cict` type containing keys and values. First up, let's look for the unique values in the `Series`.

In [36]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'opel', 'mazda',
       'porsche', 'mini', 'toyota', 'dacia', 'saab', 'volvo', 'nissan',
       'mitsubishi', 'jaguar', 'fiat', 'skoda', 'subaru',
       'sonstige_autos', 'kia', 'citroen', 'chevrolet', 'hyundai',
       'honda', 'daewoo', 'suzuki', 'land_rover', 'jeep', 'alfa_romeo',
       'rover', 'daihatsu', 'lancia', 'lada', 'trabant'], dtype=object)

In [40]:
## Do the same sorting code as the previous ones,
## you don't sort the letters and you sort the values descendingly.

brand_counts = (autos["brand"]
              .value_counts(
              normalize=True, 
              dropna=False)
             )

print(brand_counts)

volkswagen        0.204326
bmw               0.110551
opel              0.110436
mercedes_benz     0.092249
audi              0.087258
ford              0.071868
renault           0.049845
peugeot           0.031451
fiat              0.026367
seat              0.019365
skoda             0.017516
mazda             0.015598
nissan            0.015413
smart             0.015252
citroen           0.014535
toyota            0.013033
hyundai           0.010792
mini              0.009359
volvo             0.008596
mitsubishi        0.008504
kia               0.007626
honda             0.007510
sonstige_autos    0.006748
alfa_romeo        0.006586
suzuki            0.005962
porsche           0.005315
chevrolet         0.005084
chrysler          0.003513
dacia             0.002842
daihatsu          0.002657
land_rover        0.002149
jeep              0.002103
subaru            0.002080
daewoo            0.001618
saab              0.001571
jaguar            0.001433
rover             0.001387
l

As you can see in the output above, brand `volkswagen` is sold the most with a percentage of 20,4%. We can aggregate the brand that makes up >5% of the sales and see the mean price for each one. We're using a built-in function `index` to show us what brands are sold the most. Let's dive in, shall we?

In [47]:
# Define the brands we're going to aggregate:
common_brands = brand_counts[brand_counts > .05].index
print(common_brands, "\n\n")

# Define an empty dictionary
brand_mean_prices = {}

# Loop the DataFrame with each brands
for b in common_brands:
    
    # Define the selected rows as
    # the DataFrame whose `brand` is in the list.
    brand_only = autos[ autos["brand"] == b ]
    
    # USE THE BUILT-IN METHOD: `mean()`.
    # LESSON LEARNED: Look. A lot. in. the.
    # DO-CU-MEN-TA-TION! It's handy for you.
    mean_price = brand_only["price"].mean()
    
    # Put the mean in the `dict` as value,
    # also with the brand as the [key]
    brand_mean_prices[b] = round(mean_price, 3)

print(brand_mean_prices)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object') 


{'volkswagen': 5699.139, 'bmw': 8687.546, 'opel': 3015.35, 'mercedes_benz': 8922.736, 'audi': 9837.648, 'ford': 3495.034}


As we can see, there are 6 cars that makes up at least 5% of the sales. Their sales range from 3015 to 9837 dollars. And the price gap is visible. Why could it be? Is it because of the feature they give? And.. can we put both data at once we can compare it easily? Let's find out.

## Data Exploration: Brand by Average Mileage

Use the loop method from the previous one to aggregate `brand` to determine the average mileage.

In [48]:
# Define an empty dictionary
brand_avg_mile = {}

# Loop the DataFrame with each brands
for b in common_brands:
    
    # Define the selected rows as
    # the DataFrame whose `brand` is in the list.
    brand_only = autos[ autos["brand"] == b ]
    
    # USE THE BUILT-IN METHOD: `mean()`.
    # LESSON LEARNED: Look. A lot. in. the.
    # DO-CU-MEN-TA-TION! It's handy for you.
    avg_mile = brand_only["odometer_km"].mean()
    
    # Put the mean in the `dict` as value,
    # also with the brand as the [key]
    brand_avg_mile[b] = round(avg_mile, 3)

print(brand_avg_mile)

{'volkswagen': 128384.415, 'bmw': 132195.861, 'opel': 129931.994, 'mercedes_benz': 130199.148, 'audi': 128086.6, 'ford': 125411.576}


Now that we've got both data, we can make it into a new table that we can save on. How, you ask? We can learn by doing by using:
* pandas series constructor
* pandas dataframe constructor

In [56]:
# Series: Make the `dict` into a `Series`
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending=False)
print(bmp_series)

audi             9837.648
mercedes_benz    8922.736
bmw              8687.546
volkswagen       5699.139
ford             3495.034
opel             3015.350
dtype: float64


In [61]:
# Series: Make the `dict` into a `Series`
df_mp_am = pd.DataFrame(bmp_series, columns=['mean_price'])
print(df_mp_am)

               mean_price
audi             9837.648
mercedes_benz    8922.736
bmw              8687.546
volkswagen       5699.139
ford             3495.034
opel             3015.350


We can put another `Series` in a `DataFrame` to combine the data.

In [62]:
bmam_series = pd.Series(brand_avg_mile).sort_values(ascending=False)
print(bmam_series)

bmw              132195.861
mercedes_benz    130199.148
opel             129931.994
volkswagen       128384.415
audi             128086.600
ford             125411.576
dtype: float64


In [64]:
# WHen we have a made `DataFrame` already,
# we can just assign a new series as usual.

df_mp_am["avg_mile"] = bmam_series
print(df_mp_am)

               mean_price    avg_mile
audi             9837.648  128086.600
mercedes_benz    8922.736  130199.148
bmw              8687.546  132195.861
volkswagen       5699.139  128384.415
ford             3495.034  125411.576
opel             3015.350  129931.994


We can see that the correlations on both data are not that much. So it may not be the reason the car is commonly sold.

## Aaaand that's about it.
## Congratz for finishing the guided project!

In this guided project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings. Here are some next steps for you to consider:

* Data cleaning next steps:
    * Identify categorical data that uses german words, translate them and map the values to their english counterparts
    * Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
    * See if there are particular keywords in the name column that you can extract as new columns

* Analysis next steps:
    * Find the most common brand/model combinations
    * Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
    * How much cheaper are cars with damage than their non-damaged counterparts?